agent: fix accidental immediate reboots on hosts that use specialisat…

ctheune opened the pull request:
…ions

Fixes PL-133685

@flyingcircusio/release-managers

## Release process

- [x] Created changelog entry using `./changelog.sh`

## PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [x] ticket state set to *Pull request ready*
- [ ] if ticket is more urgent than within the next few days, directly contact a member of the Platform team
<!---->
- [x] set urgency and risk labels
- [ ] ensure the merge bot has determined a merge date
- [ ] ensure all checks are green
- [ ] get a review from a colleague

## Design notes

- [ ] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
- [ ] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)

general system stability

- [x] Security requirements tested? (EVIDENCE)

* manually tested on dev router
* added unit test coverage

ctheune committed:
agent: fix accidental immediate reboots on hosts that use specialisations

Fixes PL-133685

osnyx committed:
agent: extend channel reboot detection tests

Our interactive tooling `fc-manage switch -s "somespecialisation"` also
supports switching to a new system on a certain specialisation. The
reboot detection shall only trigger for numeric release changes for
these cases as well.

osnyx committed:
agent: fc.util.nixos.get_release_version strips prefixes

Necessary for properly comparing different specialisations within the
same NixOS release series.

osnyx reviewed (dismissed):
LGTM what was there, this specifically fits the issue we encountered.

In general though, the specialisations can possibly change at the same time as well as the NixOS version, at least our `fc-manage switch -s somespecialisation` supports that. So I extended the test cases to cover this and made some necessary adjustments.

This now needs your review.

ctheune committed:
agent: consistently switch to pyproject and use local `uv` workflow

(also causes reformat)

ctheune committed:
agent: rework current/next version detection

Do not use complicated parsing in `nixos-version` but leverage the
standardizes `os-release` format.

osnyx commented:
@ctheune Interesting, do you have different pre-commit tooling (especially isort) than the CI has? https://github.com/flyingcircusio/fc-nixos/actions/runs/15007671114/job/42170173821?pr=1470

ctheune commented:
> @ctheune Interesting, do you have different pre-commit tooling (especially isort) than the CI has? https://github.com/flyingcircusio/fc-nixos/actions/runs/15007671114/job/42170173821?pr=1470

No. The issue here is that the effect of isort differs between it being run within a git pre-commit hook and from CLI. Trying to figure out what's going on here. The same thing happens locally when I run `pre-commit run  --all-files`. It creates a battle with the git-invoked hook.

ctheune reviewed (commented):
LGTM (weird comment before, because the approval button was confusing me)
